File: src/draw-v1/lib/color_wheel.py
import math

from PySide6.QtCore import Signal, QPoint, QRectF, Qt
from PySide6.QtGui import QColor, QPainter, QPen, QMouseEvent
from PySide6.QtWidgets import QWidget


class ColorWheel(QWidget):
    color_changed = Signal(QColor)

    def __init__(self, circle_radius=45, square_size=40, thickness=13):
        super().__init__()
        self.setMinimumSize(circle_radius * 2, circle_radius * 2)
        self.hue = 0
        self.saturation = 1.0
        self.value = 1.0
        self.radius = circle_radius
        self.sq_size = square_size
        self.thickness = thickness

        self.is_hue_wheel = None

    def get_color(self):
        return QColor.fromHsv(self.hue, int(self.saturation * 255), int(self.value * 255))

    def paintEvent(self, event):
        painter = QPainter(self)
        painter.setRenderHint(QPainter.RenderHint.Antialiasing)
        self.radius = min(self.width(), self.height()) // 2 - 10
        center = QPoint(self.width() // 2, self.height() // 2)

        # ring
        for angle in range(360):
            color = QColor.fromHsv(angle, 255, 255)
            painter.setPen(QPen(color, self.thickness))
            painter.drawArc(QRectF(center.x() - self.radius, center.y() - self.radius,
                                   self.radius * 2, self.radius * 2), angle * 16, 16)

        # square
        square_size = self.sq_size
        square_top_left = QPoint(center.x() - square_size // 2, center.y() - square_size // 2)
        for x in range(int(square_size)):
            for y in range(int(square_size)):
                s = x / square_size
                v = 1 - y / square_size
                color = QColor.fromHsv(self.hue, int(s * 255), int(v * 255))
                painter.setPen(color)
                painter.drawPoint(square_top_left + QPoint(x, y))

        painter.setPen(QPen(Qt.GlobalColor.white, 2))
        painter.setBrush(Qt.BrushStyle.NoBrush)

        # Hue
        angle_rad = math.radians(self.hue)
        hx = center.x() + self.radius * math.cos(angle_rad)
        hy = center.y() - self.radius * math.sin(angle_rad)
        painter.drawEllipse(QPoint(hx, hy), 5, 5)

        # SV
        sv_x = int(square_top_left.x() + self.saturation * square_size)
        sv_y = int(square_top_left.y() + (1 - self.value) * square_size)
        painter.drawEllipse(QPoint(sv_x, sv_y), 5, 5)

    def mousePressEvent(self, event: QMouseEvent):
        center = QPoint(self.width() // 2, self.height() // 2)
        dx = event.position().x() - center.x()
        dy = event.position().y() - center.y()
        dist = math.hypot(dx, dy)
        if self.radius - 10 < dist < self.radius + 10:
            self.is_hue_wheel = True
        else:
            self.is_hue_wheel = False

        self.handle_mouse(event)

    def mouseMoveEvent(self, event: QMouseEvent):
        self.handle_mouse(event)

    def handle_mouse(self, event: QMouseEvent):
        center = QPoint(self.width() // 2, self.height() // 2)
        dx = event.position().x() - center.x()
        dy = event.position().y() - center.y()

        if self.is_hue_wheel:
            # Hue ring
            angle = math.degrees(math.atan2(-dy, dx)) % 360
            self.hue = int(angle)
            self.update()
        else:
            # SV square
            top_l = QPoint(center.x() - self.sq_size // 2, center.y() - self.sq_size // 2)
            x = event.position().x() - top_l.x()
            y = event.position().y() - top_l.y()
            self.saturation = min(max(x / self.sq_size, 0), 1)
            self.value = 1 - min(max(y / self.sq_size, 0), 1)
            self.update()

        self.color_changed.emit(self.get_color())

File: src/Quol-CPP/plugins/draw/lib/ColorWheel.cpp
#include "plugins/draw/lib/ColorWheel.hpp"

#include <QImage>
#include <QMouseEvent>
#include <QPainter>
#include <QPen>
#include <algorithm>
#include <cmath>

ColorWheel::ColorWheel(int circleRadius, int squareSize, int thickness, QWidget *parent)
    : QWidget(parent), m_circleRadius(circleRadius), m_thickness(thickness) {
    // Fit the SV square inside the inner edge of the ring
    const int innerRadius = m_circleRadius - m_thickness;
    const int autoSquare = static_cast<int>(innerRadius / 1.2);
    m_squareHalf = (squareSize > 0) ? squareSize / 2 : autoSquare;
    const int side = circleRadius * 2 + 16;
    setFixedSize(side, side);
    setSizePolicy(QSizePolicy::Fixed, QSizePolicy::Fixed);
    setCursor(Qt::CrossCursor);
}

QColor ColorWheel::color() const {
    return QColor::fromHsv(m_hue, qRound(m_saturation * 255), qRound(m_value * 255));
}

void ColorWheel::setColor(const QColor &c, bool emitSignal) {
    if (!c.isValid())
        return;
    int h, s, v, a;
    c.getHsv(&h, &s, &v, &a);
    if (h != -1)
        m_hue = h;  // -1 for achromatic colours
    m_saturation = s / 255.0;
    m_value = v / 255.0;
    update();
    if (emitSignal)
        emit colorChanged(color());
}

// ---------------------------------------------------------------------------
// Paint
// ---------------------------------------------------------------------------

void ColorWheel::paintEvent(QPaintEvent *) {
    QPainter p(this);
    p.setRenderHint(QPainter::Antialiasing);

    const QPoint c = center();
    const int r = ringRadius();
    const int sh = squareHalf();
    const QRectF ring(c.x() - r, c.y() - r, r * 2, r * 2);

    // ── Hue ring ──────────────────────────────────────────────────────────
    QPen pen;
    pen.setWidth(m_thickness);
    for (int deg = 0; deg < 360; ++deg) {
        pen.setColor(QColor::fromHsv(deg, 255, 255));
        p.setPen(pen);
        // Qt arc: angle in 1/16 degree, counter-clockwise, 0 = 3 o'clock
        p.drawArc(ring, deg * 16, 16);
    }

    // ── SV square ─────────────────────────────────────────────────────────
    // Build with a QImage for pixel accuracy
    const int sz = sh * 2;
    if (sz > 0) {
        QImage img(sz, sz, QImage::Format_RGB32);
        for (int x = 0; x < sz; ++x) {
            for (int y = 0; y < sz; ++y) {
                double s = double(x) / sz;
                double v = 1.0 - double(y) / sz;
                img.setPixel(x, y, QColor::fromHsv(m_hue, qRound(s * 255), qRound(v * 255)).rgb());
            }
        }
        p.drawImage(c.x() - sh, c.y() - sh, img);
    }

    // ── Indicators ────────────────────────────────────────────────────────
    p.setBrush(Qt::NoBrush);

    // Hue indicator on ring
    {
        const double rad = qDegreesToRadians(double(m_hue));
        const int ix = qRound(c.x() + r * std::cos(rad));
        const int iy = qRound(c.y() - r * std::sin(rad));
        p.setPen(QPen(Qt::white, 2));
        p.drawEllipse(QPoint(ix, iy), 5, 5);
    }

    // SV indicator inside square
    {
        const int ix = qRound(c.x() - sh + m_saturation * sz);
        const int iy = qRound(c.y() - sh + (1.0 - m_value) * sz);
        p.setPen(QPen(Qt::white, 2));
        p.drawEllipse(QPoint(ix, iy), 5, 5);
    }
}

// ---------------------------------------------------------------------------
// Mouse
// ---------------------------------------------------------------------------

void ColorWheel::mousePressEvent(QMouseEvent *e) {
    if (e->button() != Qt::LeftButton)
        return;

    const QPoint c = center();
    const double dx = e->position().x() - c.x();
    const double dy = e->position().y() - c.y();
    const double dist = std::hypot(dx, dy);
    const int r = ringRadius();

    // Decide whether dragging the ring or the square
    m_onRing = (dist >= r - m_thickness - 2 && dist <= r + 4);
    m_dragging = true;
    handleMouse(e->position());
}

void ColorWheel::mouseMoveEvent(QMouseEvent *e) {
    if (m_dragging)
        handleMouse(e->position());
}

void ColorWheel::mouseReleaseEvent(QMouseEvent *e) {
    if (e->button() == Qt::LeftButton)
        m_dragging = false;
}

void ColorWheel::handleMouse(const QPointF &pos) {
    const QPoint c = center();
    const double dx = pos.x() - c.x();
    const double dy = pos.y() - c.y();

    if (m_onRing) {
        double angle = std::fmod(qRadiansToDegrees(std::atan2(-dy, dx)), 360.0);
        if (angle < 0)
            angle += 360.0;
        m_hue = qRound(angle);
    } else {
        const int sh = squareHalf();
        const int sz = sh * 2;
        if (sz <= 0)
            return;
        double x = pos.x() - (c.x() - sh);
        double y = pos.y() - (c.y() - sh);
        m_saturation = qBound(0.0, x / sz, 1.0);
        m_value = qBound(0.0, 1.0 - y / sz, 1.0);
    }

    update();
    emitColor();
}

void ColorWheel::emitColor() {
    emit colorChanged(color());
}
